src/transformStateWithClones.js

```diff
@@ -7,7 +7,26 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  const copiedState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      Object.assign(copiedState, action.extraData);
+      stateHistory.push(Object.assign({}, copiedState));
+    } else if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete copiedState[key];
+      }
+
+      stateHistory.push(Object.assign({}, copiedState));
```

Comment:
The `stateHistory.push` operation should be moved outside of the `if-else` block and placed at the end of each loop cycle. This ensures that the state is consistently recorded after each action, as per the additional prompt rules.